scour/scour.py

```diff
@@ -1355,6 +1355,32 @@ def collapseSinglyReferencedGradients(doc):
     return num
 
 
+def computeGradientBucketKey(grad):
+    # Compute a key (hashable opaque value; here a string) from each
+    # gradient such that "key(grad1) == key(grad2)" is the same as
+    # saying that grad1 is a duplicate of grad2.
+    gradBucketAttr = ['gradientUnits', 'spreadMethod', 'gradientTransform',
+                      'x1', 'y1', 'x2', 'y2', 'cx', 'cy', 'fx', 'fy', 'r']
+    gradStopBucketsAttr = ['offset', 'stop-color', 'stop-opacity', 'style']
+
+    # A linearGradient can never be a duplicate of a
+    # radialGradient (and vice versa)
+    subKeys = [grad.getAttribute(a) for a in gradBucketAttr]
+    subKeys.append(grad.getAttributeNS(NS['XLINK'], 'href'))
+    stops = grad.getElementsByTagName('stop')
+    if stops.length:
+        for i in range(stops.length):
+            stop = stops.item(i)
+            for attr in gradStopBucketsAttr:
+                stopKey = stop.getAttribute(attr)
+                subKeys.append(stopKey)
```

Comment:
Can we guarantee this code will always produce unique keys?

E.g. when considering different number of stops / different attributes set? Especially with default attributes we produce a lot of repeating `\x1e\x1e`.

Reply:
The number of elements per gradient and per stop is always constant as the append-operation is not conditional. They only case, where there might be a problem is if "attribute is defined explicitly to be empty" is different from "attribute is not defined at all" (both of those seem to make getAttribute return the empty string).  However, the original code did not distinguish between those two states, so I assumed I could rely on that as well.

That said, it should be a simple matter to use a different separator between each stop and a sentential value for "empty"/"undefined" attributes.